app/callbacks/views/sns.py

```diff
@@ -0,0 +1,416 @@
+                    sys.exc_info()[1]
+                )
+            ),
+            log_level=logging.INFO,
+            condition=True,
+        ) as log_context:
+            confirmation_url = body_json["SubscribeURL"]
+            log_context.update({
+                "target_url": confirmation_url,
+                "topic_arn": topic_arn,
+            })
+            confirmation_response = requests.get(confirmation_url)
+            confirmation_response.raise_for_status()
+    except requests.exceptions.RequestException:
+        abort(400, f"SubscriptionConfirmation request to {confirmation_url!r} failed")
+
+    # check we received what we expected from url
+    subscription_arn = confirmation_request_id = None
+    try:
+        confirmation_etree = etree_fromstring(confirmation_response.content)
+        namespaces = {"n": confirmation_etree.nsmap[None]}
+        subscription_arn = confirmation_etree.xpath(
+            "normalize-space(string(/n:ConfirmSubscriptionResponse/n:ConfirmSubscriptionResult/n:SubscriptionArn))",
+            namespaces=namespaces,
+        )
+        confirmation_request_id = confirmation_etree.xpath(
+            "normalize-space(string(/n:ConfirmSubscriptionResponse/n:ResponseMetadata/n:RequestId))",
+            namespaces=namespaces,
+        )
+    except ParseError as e:
+        current_app.logger.warning(e)
+
+    if not subscription_arn:
+        current_app.logger.warning("SubscriptionConfirmation response parsing failed")
+        abort(400, f"SubscriptionConfirmation request to {confirmation_url!r}: response parsing failed")
+    else:
+        current_app.logger.info(
+            "SubscriptionConfirmation succeeded for subscription {subscription_arn}",
+            extra={
+                "subscription_arn": subscription_arn,
+                "confirmation_request_id": confirmation_request_id,
+            },
+        )
+
+    return jsonify(status='ok'), 200
+
+
+@lru_cache()
+def _get_certificate(url):
+    with logged_duration(
+        logger=current_app.logger,
+        message=lambda _: (
+            "Fetched certificate for SNS signature validation from {target_url}"
+            if sys.exc_info()[0] is None else
+            # need to literally format() the exception into the message as it's difficult to get it injected into extra
+            "Failed to fetch certificate for SNS signature validation from {{target_url}}: {!r}".format(
+                sys.exc_info()[1]
+            )
+        ),
+        log_level=logging.INFO,
+        condition=True,
+    ) as log_context:
+        log_context.update({"target_url": url})
+        response = requests.get(url)
+        response.raise_for_status()
+        return response.content
+
+
+# deliberately omitting the optional trailing .cn here
+VALIDATION_CERTIFICATE_URL_REGEX = r"^https://[-a-z0-9.]+\.amazonaws\.com/"
+# give request extra headroom of a minute here
+VALIDATION_MAX_AGE = datetime.timedelta(hours=1, minutes=1)
```

Comment:
y?

Reply:
Just using my :crystal_ball: to predict obvious unnecessary failures that might occur and annoy us.

Reply:
It's gone from 1 hour, to 1 hour and 1 minute?

Reply:
:grin: